testsuite report: the "list of tests taking more than 10s" is taking too long to skip

When I see a CI failure in the github CI, my usual workflow to tell which test failed is as follows:
- pick one of the failed tests, click on it
- I get brought to a page with an annoying web UI around the test logs, scroll down to the end
- read what is at the end to tell which test is failing
- sometimes, scroll back up to get that test's output (this can usually be done by using "search" on the tests name)

On a `macos` log today, what I see after scrolling at the end is as follows:

```
    tests/parallel/backup_thread_pipe.ml: 10.21
    tests/parallel/fib_threads.ml: 65.83
    tests/parallel/mctest.ml: 11.71
    tests/regression/pr5757/pr5757.ml: 27.82
    tests/regression/pr9853/compaction_corner_case.ml: 15.17
    tests/unboxed-primitive-args/test.ml: 12.96

#### Something failed. Exiting with error status.

make[1]: *** [report] Error 4
make: *** [parallel-*] Error 2
Error: Process completed with exit code 2.
```

None of the test names listed here have the information I am looking for, they are just noise from my perspective.

This is because the structure of the test summary is as follows:

```
List of skipped tests:
[58 lines of output]

List of unexpected errors:
    tests/typing-signatures/regression_tsubst_error.ml

Summary:
  3354 tests passed
   64 tests skipped
    0 tests failed
  123 tests not started (parent test skipped or failed)
    1 unexpected errors
  3542 tests considered

Tests taking longer than 10s:
[25 lines of output]
```

My proposal: include the "Tests taking longer than 10s" list right after the "List of skipped tests" list, so that I can easily ignore them both and get the important information first.

--- a/testsuite/summarize.awk
+++ b/testsuite/summarize.awk
@@ -211,13 +211,13 @@ END {
         }
         printf("\n");
         printf("Summary:\n");
-        printf("  %3d tests passed\n", passed);
-        printf("  %3d tests skipped\n", skipped);
-        printf("  %3d tests failed\n", failed);
-        printf("  %3d tests not started (parent test skipped or failed)\n",
+        printf("  %4d tests passed\n", passed);
+        printf("  %4d tests skipped\n", skipped);
+        printf("  %4d tests failed\n", failed);
+        printf("  %4d tests not started (parent test skipped or failed)\n",
                ignored);
-        printf("  %3d unexpected errors\n", unexped);
-        printf("  %3d tests considered", nresults);
+        printf("  %4d unexpected errors\n", unexped);
+        printf("  %4d tests considered", nresults);
         if (nresults != passed + skipped + ignored + failed + unexped){
             printf (" (totals don't add up??)");
         }

--- a/testsuite/summarize.awk
+++ b/testsuite/summarize.awk
@@ -13,6 +13,14 @@
 #*                                                                        *
 #**************************************************************************
 
+BEGIN {
+  # awk trick to detect whether a function exists. x is undefined. If asort is
+  # present then asort (x) returns 0 (no elements sorted). If asort is not
+  # present then the _space_ before the bracket turns it into string
+  # concatenation of two undefined variables (which returns "").
+  has_asort = (asort (x) == "0")
+}
+
 function check() {
     if (!in_test){
         printf("error at line %d: found test result without test start\n", NR);
@@ -176,6 +184,13 @@ END {
             }else if (r == "n"){
                 ++ ignored;
             }
+        }
+        if (has_asort) {
+          asort(skips);
+          asort(blanks);
+          asort(fail);
+          asort(unexp);
+          asort(slow);
         }
         printf("\n");
         if (skipped != 0){

--- a/testsuite/summarize.awk
+++ b/testsuite/summarize.awk
@@ -152,6 +152,7 @@ function record_unexp() {
 }
 
 END {
+    if (ENVIRON["GITHUB_ACTIONS"] == "true") start_group = "::group::";
     if (in_test) record_unexp();
 
     if (errored){
@@ -197,12 +198,14 @@ END {
         isort(slow, slowcount);
         printf("\n");
         if (skipped != 0){
-            printf("\nList of skipped tests:\n");
+            printf("\n%sList of skipped tests:\n", start_group);
             for (i=0; i < skipidx; i++) printf("    %s\n", skips[i]);
+            if (ENVIRON["GITHUB_ACTIONS"] == "true") print "::endgroup::";
         }
         if (slowcount != 0){
-            printf("\n\nTests taking longer than 10s:\n");
+            printf("\n\n%sTests taking longer than 10s:\n", start_group);
             for (i=0; i < slowcount; i++) printf("    %s\n", slow[i]);
+            if (ENVIRON["GITHUB_ACTIONS"] == "true") print "::endgroup::";
         }
         if (empty != 0){
             printf("\nList of directories returning no results:\n");

--- a/testsuite/summarize.awk
+++ b/testsuite/summarize.awk
@@ -13,12 +13,17 @@
 #*                                                                        *
 #**************************************************************************
 
-BEGIN {
-  # awk trick to detect whether a function exists. x is undefined. If asort is
-  # present then asort (x) returns 0 (no elements sorted). If asort is not
-  # present then the _space_ before the bracket turns it into string
-  # concatenation of two undefined variables (which returns "").
-  has_asort = (asort (x) == "0")
+# Insertion sort in awk, because asort is a GNU extension. Lifted carefully
+# with tongs out of the mawk manpage.
+function isort(A, n) {
+  for (i = 1; i < n; i++) {
+    hold = A[j = i];
+    while (A[j-1] > hold) {
+      j--;
+      A[j+1] = A[j];
+    }
+    A[j] = hold;
+  }
 }
 
 function check() {
@@ -185,13 +190,11 @@ END {
                 ++ ignored;
             }
         }
-        if (has_asort) {
-          asort(skips);
-          asort(blanks);
-          asort(fail);
-          asort(unexp);
-          asort(slow);
-        }
+        isort(skips, skipidx);
+        isort(blanks, empty);
+        isort(fail, failed);
+        isort(unexp, unexped);
+        isort(slow, slowcount);
         printf("\n");
         if (skipped != 0){
             printf("\nList of skipped tests:\n");

--- a/testsuite/summarize.awk
+++ b/testsuite/summarize.awk
@@ -197,6 +197,10 @@ END {
             printf("\nList of skipped tests:\n");
             for (i=0; i < skipidx; i++) printf("    %s\n", skips[i]);
         }
+        if (slowcount != 0){
+            printf("\n\nTests taking longer than 10s:\n");
+            for (i=0; i < slowcount; i++) printf("    %s\n", slow[i]);
+        }
         if (empty != 0){
             printf("\nList of directories returning no results:\n");
             for (i=0; i < empty; i++) printf("    %s\n", blanks[i]);
@@ -220,10 +224,6 @@ END {
         printf("  %4d tests considered", nresults);
         if (nresults != passed + skipped + ignored + failed + unexped){
             printf (" (totals don't add up??)");
-        }
-        if (slowcount != 0){
-            printf("\n\nTests taking longer than 10s:\n");
-            for (i=0; i < slowcount; i++) printf("    %s\n", slow[i]);
         }
         printf ("\n");
         if (failed || unexped){

